Reference example for gRPC java client (#1236)
stargate/stargate: grpc/README.md

`No functional channel service provider found. Try adding a dependency on the grpc-okhttp, grpc-netty, or grpc-netty-shaded artifact`.

Once we have all needed dependencies we should be able to use the Stargate gRPC-stub API. 
After this step, you should have `StargateGrpc` available on your class path (from the `grpc-proto` dependency).
It contains the logic for interacting with Stargate gRPC API.

### Usage

Once we have the generated code, we are ready to create the client based on that.
Before delving into the code, we need to generate the auth token that will be used to perform Authorization, 
to see how to do it, please visit the [Stargate Authz documentation].
For the development purpose, this should be good enough:
```shell script
curl -X POST localhost:8081/v1/auth/token/generate --header "Content-Type: application/json" --data '{"key":"cassandra","secret":"cassandra"}'
```
assuming that Stargate is running on the `localhost:8081`  

Once we have that, we can connect to the gRPC API. Firstly, we need to generate the `Channel` that is used to perform connection:
```java
public ManagedChannel createChannel(String host, int port) {

>     return ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();

Reviewer: maybe add a small note that `usePlaintext()` should only be used for testing and non-sensitive data? Actually removing it could be even better?

Author: @mpenick what's the recommended way for clients to create this channel?
When I drop `usePlaintext()`, I got:
```
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
Channel Pipeline: [SslHandler#0, ProtocolNegotiators$ClientTlsHandler#0, WriteBufferingAndExceptionHandler#0, DefaultChannelPipeline$TailContext#0]
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:262)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:243)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:156)
	at io.stargate.proto.StargateGrpc$StargateBlockingStub.executeQuery(StargateGrpc.java:236)
	at GrpcClient.executeQuery(GrpcClient.java:27)
	at GrpcClient.main(GrpcClient.java:61)
```
locally

Reviewer 2: Ivan is correct, `usePlaintext()` should only be used for development and testing. When using in a production environment it should use a load balancer that terminates TLS or the gRPC needs to support TLS (which it currently doesn't).

Author: Do you have an example from the client perspective how this should be configured?

Author: I mentioned it on:
https://github.com/stargate/stargate/pull/1236/commits/665ce94e4bb88f0d415fd113c93a9e491fbfc9f2

# gRPC Service

## gRPC Java Client 

### Setup

Let's see how to configure a java project to use the Stargate gRPC API.
The generated code based on protobuf files (`query.proto` and `stargate.proto`) is shipped with the `grpc-proto` dependency.

> To see a guide how it compiles the Java code see: [gRPC setup project dependencies].

Reviewer 3: ```suggestion
To see a guide how the Java code is compiled from the proto files see: [gRPC setup project dependencies]. This is purely background information and not required for using the client discussed here.
```

When I first clicked the link I was a little confused about how it applies to the client.

> In your client application, you only need to add one dependency:
> 
> ```xml
> <dependencies>
>     <dependency>
>         <groupId>io.stargate.grpc</groupId>
>         <artifactId>grpc-proto</artifactId>
>         <version>1.0.32</version>
>     </dependency>
> </dependencies>
> ```
> The last missing piece is to add a functional channel service provider. We pick netty:
> ```xml
> <dependencies>
>     <dependency>
>             <groupId>io.grpc</groupId>
>             <artifactId>grpc-netty-shaded</artifactId>
>             <version>1.40.1</version>
>     </dependency>
> </dependencies>
> ```

Reviewer 3: ```suggestion
In your client application, you only need to add two dependencies which are the client and a functional channel service provider (we pick netty here).

```xml
<dependencies>
    <dependency>
        <groupId>io.stargate.grpc</groupId>
        <artifactId>grpc-proto</artifactId>
        <version>1.0.32</version>
    </dependency>
    <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-netty-shaded</artifactId>
            <version>1.40.1</version>
    </dependency>
</dependencies>
```
```

After this step, you should have `StargateGrpc` available on your class path (from the `grpc-proto` dependency).
It contains the logic for interacting with Stargate gRPC API.

### Usage

Once we have the generated code, we are ready to create the client based on that.
Before delving into the code, we need to generate the auth token that will be used to perform Authorization, 
to see how to do it, please visit the [Stargate Authz documentation].
For the development purpose, this should be good enough:
```shell script
curl -X POST localhost:8081/v1/auth/token/generate --header "Content-Type: application/json" --data '{"key":"cassandra","secret":"cassandra"}'
```
assuming that Stargate is running on the `localhost:8081`  

Once we have that, we can connect to the gRPC API. Firstly, we need to generate the `Channel` that is used to perform connection:
```java
public ManagedChannel createChannel(String host, int port) {
    return ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
}
```
Please note that `usePlaintext()` should only be used for development and testing.
When using in a production environment it should use a load balancer that terminates TLS.

> For local development of stargate, it will be:

Reviewer 3: ```suggestion
For local development of Stargate, it will be:
```

```java
ManagedChannel channel = createChannel("localhost", 8090);
```

Next, we can generate the `StargateGrpc` stub. There are two ways of interacting with the gRPC API.

The first one is synchronous (blocking):
```java
import io.stargate.grpc.StargateBearerToken;

StargateGrpc.StargateBlockingStub blockingStub = StargateGrpc.newBlockingStub(channel)
                                                    .withCallCredentials(new StargateBearerToken("token-value"))
                                                    .withDeadlineAfter(5, TimeUnit.SECONDS);
```

The second one is async (non-blocking):
```java

> StargateGrpc.StargateStub = = StargateGrpc.newStub(channel)

Reviewer 2: `= =`

``` 

The Stargate gRPC API provides a way to execute two types of queries:
- standard CQL queries
- batch queries (contains N CQL queries)

#### Standard Query

We can start from inserting a record using gRPC stub:

```java
import io.stargate.proto.QueryOuterClass.Response;
import io.stargate.proto.QueryOuterClass;

Response response =
        blockingStub.executeQuery(
                        QueryOuterClass.Query
                                    .newBuilder().setCql("INSERT INTO ks.test (k, v) VALUES ('a', 1)").build());
```
This will build and execute a single query. 

Next, we can retrieve the inserted record(s):
```java
Response response = stub.executeQuery(QueryOuterClass.Query.newBuilder().setCql("SELECT k, v FROM ks.test").build());

```

If we print out the result set, it will have the following structure:
```yaml
result_set {
  data {
    type_url: "type.googleapis.com/stargate.ResultSet"
    value: "some_binary_data"
  }
}
```    
The value contains the binary data, that we can deserialize. 
Firstly, we need to unwrap the `ResultSet`:
```java
ResultSet rs = response.getResultSet().getData().unpack(QueryOuterClass.ResultSet.class);

> rs.getRows(0).getValues(0).getString(); // it will return value for k = "a"

Reviewer: what about if we replace this with the example of get the row list and iterating over, some kind of example that is not so bounded on the type of the query, would this be possible? starter programmers could like this better I guess..

Author: done

rs.getRows(0).getValues(1).getInt(); // it will return value for v = 1             
```
We can get all rows `getRowsList()` and iterate over the result 

> OR get the specific row using its index and passing it to the `getRows(int index)` method. We picked the formed approach.

Reviewer 3: ```suggestion
OR get the specific row using its index and passing it to the `getRows(int index)` method. We picked the latter approach in the example above.
```

> Our retrieval query (`SELECT k, v FROM test"`) stated explicitly which columns should be retrieved.

Reviewer 3: ```suggestion
Our retrieval query (`SELECT k, v FROM ks.test"`) stated explicitly which columns should be retrieved.
```

Thanks to that, we can safely get the values using their positions via the `getValues()` method.

> The `getString()` and `getInt()` perform deserialization of data. The API provides utility methods for deserialization for more types.

Reviewer 3: ```suggestion
The `getString()` and `getInt()` perform deserialization of data. These methods were used because we knew the underlying type of the corresponding columns. The API provides utility methods for deserialization for more types as well.
```

Clarifying for the "Well why did you use getString() here instead of X?" question.

For the full list of available types, see `Value` section in the [query.proto] file. 
 
#### Batch Query

In case we want to execute N queries, we can use the `executeBatch` method:
```java
     QueryOuterClass.Response response =
                blockingStub.executeBatch(
                        QueryOuterClass.Batch.newBuilder()
                                .addQueries(QueryOuterClass.BatchQuery.newBuilder().setCql("INSERT INTO ks.test (k, v) VALUES ('a', 1)").build())
                                .addQueries(
                                        QueryOuterClass.BatchQuery.newBuilder().setCql("INSERT INTO ks.test (k, v) VALUES ('b', 2)").build())
                                .build());

```

> It takes the `Batch` as an argument. A Batch can contain N queries. We are adding two queries via `addQueries` method.

Reviewer 3: ```suggestion
It takes the `Batch` as an argument. A Batch can contain N queries of type INSERT, UPDATE, or DELETE. We are adding two queries via `addQueries` method.
```

Might also want to consider linking to the batch docs somewhere
https://docs.datastax.com/en/dse/6.0/cql/cql/cql_reference/cql_commands/cqlBatch.html

#### Async API

Up to this point, we were using the blocking version of the generated stub. 
We can also interact with the Stargate API using the async version of the stub.
To do so, we need to pass the [StreamObserver] that will be called asynchronously when the results are available.

Every StreamObserver needs to implement 3 methods: `onNext()`, `onError()` and `onComplete()`.
For example:
```java
StreamObserver<QueryOuterClass.Response> streamObserver = new StreamObserver<QueryOuterClass.Response>() {
           @Override
           public void onNext(QueryOuterClass.Response response) {
               try {
                   System.out.println("response:" + response.getResultSet().getData().unpack(QueryOuterClass.ResultSet.class));
               } catch (InvalidProtocolBufferException e) {
                   throw new RuntimeException(e);
               }
           }
           @Override
           public void onError(Throwable throwable) {
               System.out.println("Error: " + throwable);
           }
           @Override
           public void onCompleted() {
               System.out.println("completed");

>            }

Reviewer 3: Maybe pseudo code as comments or suggestions on what should be done here?

       };
```
Please note that this is a very simplified version only for demonstration purposes and should not be used on production.

Once we have the Observer, we can pass it to the `executeQuery` method on the async stub:
```java
stub.executeQuery(QueryOuterClass.Query.newBuilder().setCql("SELECT k, v FROM ks.test").build(), streamObserver);
```
This query will return immediately because it is non-blocking. 
If your program (or test) is progressing to the end, you may not be able to see the results. 

> Your program may exist before the will arrives. 

Reviewer 3: ```suggestion
Your program may exit before the data arrives. 
```
